rmw_zenoh_cpp/src/impl/wait_impl.cpp

```diff
@@ -11,15 +11,36 @@ bool check_wait_conditions(
   const rmw_guard_conditions_t * guard_conditions,
   const rmw_services_t * services,
   const rmw_clients_t * clients,
-  const rmw_events_t * events)
+  const rmw_events_t * events,
+  bool finalize)
 {
-  // Subscriptions: If there are subscription messages ready, continue
-  if (!rmw_subscription_data_t::zn_messages_.empty()) {
-    RCUTILS_LOG_DEBUG_NAMED("rmw_zenoh_cpp", "[rmw_wait] SUBSCRIPTION MESSAGES IN QUEUE: %ld",
-                             rmw_subscription_data_t::zn_messages_.size());
-    return true;
+  bool stop_wait = false;
+
+  // Subscriptions
+  if (subscriptions) {
+    size_t subscriptions_ready = 0;
+
+    for (size_t i = 0; i < subscriptions->subscriber_count; ++i) {
+        auto subscription_data = static_cast<rmw_subscription_data_t *>(subscriptions->subscribers[i]);
+        if (subscription_data->zn_message_queue_.empty()) {
+          if (finalize) {
```

Comment:
In rmw_wait, this check function is used as a predicate to wait on a condition variable.

But rcl expects rmw to set the subscription pointer it passes into rmw_wait to `NULL` if the subscription is not ready (when there is no message in the queue.)

So finalize is used to make sure that this setting to `NULL` (or in this case, `nullptr`) only ONCE. Otherwise on repeat calls to check the predicate, things will break since it'll try to compare or dereference a `nullptr`.

A call to `check_wait_conditions` with finalize set to true is guaranteed [here](https://github.com/osrf/rmw_zenoh/pull/27/files/9df834eedbc1755c39178fab5a461e272949728c)